### PR TITLE
rhncfg works in python2/3

### PR DESCRIPTION
**Pull request by Pajinek**
The dependency on rhnlib (python2) was removed for Fedora 23 and Python 3

### CODE REVIEW

File: client/tools/rhncfg/actions/ModeController.py [hunk 1]
```diff
@@ -156,7 +155,7 @@ def ask_before_continuing(question=None):
 
     while answer != 'y' and answer != 'n':
         answer = raw_input(the_question)
```
Comment (galtsib) [inline client/tools/rhncfg/actions/ModeController.py]: python 3 doesn't have raw_input()

File: client/tools/rhncfg/actions/Modes.py [hunk 2]
```diff
@@ -57,7 +57,7 @@ def set_rhn_root(self, rhn_root):
     #Creates the self.rhn_root directories if they don't already exist. This allows subclasses to implement modes in different locations.
     def _create_rhnconfig_path(self):
         if not os.path.exists(self.rhn_root):
-            os.makedirs(self.rhn_root, 0770)
+            os.makedirs(self.rhn_root, int(0o770, 8))
```
Comment (galtsib) [inline client/tools/rhncfg/actions/Modes.py]: os.makedirs(self.rhn_root, int('770', 8))

File: client/tools/rhncfg/config_common/file_utils.py [hunk 3]
```diff
@@ -215,7 +217,7 @@ def diff(self, file_struct):
                     result = ''
                 else:
                     result = "Link targets differ for [%s]: actual: [%s], expected: [%s]\n" % (path, curlink, newlink)
-            except OSError, e:
+            except OSError(e):
```
Comment (galtsib) [inline client/tools/rhncfg/config_common/file_utils.py]: why not 'except OSError as e:'?

Comment (Pajinek) [inline client/tools/rhncfg/config_common/file_utils.py]: It looks as mistake :-o

File: client/tools/rhncfg/config_common/repository.py [hunk 4]
```diff
@@ -96,7 +102,7 @@ def make_stat_info(self, path, file_stat):
             ret[label] = file_stat[st]
 
         # server expects things like 644, 700, etc.
-        ret['mode'] = deci_to_octal(ret['mode'] & 07777)
+        ret['mode'] = deci_to_octal(ret['mode'] & int(0o7777))
```
Comment (galtsib) [inline client/tools/rhncfg/config_common/repository.py]: Please, can you explain, how does it work? (for me, it looks like mistake)

File: client/tools/rhncfg/config_common/repository.py [hunk 5]
```diff
@@ -19,14 +19,20 @@
...
+from config_common import local_config
+from config_common import utils
+from config_common.rhn_log import log_debug
 
-from rhn_log import log_debug
+try: # python2
+    import xmlrpclib
+except ImportError: # python3
+    import xmlrpc.client as xmlrpclib
+    basestring = str
```
Comment (galtsib) [inline client/tools/rhncfg/config_common/repository.py]: again, if it's possible, let just use 'str'

File: client/tools/rhncfg/config_common/rpc_wrapper.py [hunk 6]
```diff
@@ -106,7 +111,7 @@ def init_server(self, myuri):
         import urllib
         self._uri = myuri
         typ, uri = urllib.splittype(self._uri)
```
Comment (galtsib) [inline client/tools/rhncfg/config_common/rpc_wrapper.py]: will not work for Python3. we need something like:

typ, uri = urllib.parse.splittype(self._uri) 

self._host, self._handler = urllib.parse.splithost(uri)

File: client/tools/rhncfg/config_common/transactions.py [hunk 7]
```diff
@@ -81,10 +81,10 @@ def _rename_to_backup(self, path):
                 # need to make sure to handle it if we catch a 'OSError: [Errno 18] Invalid cross-device link'
                 try:
                     log_debug(9, "trying to use os.renames")
-                    oumask = os.umask(022)
+                    oumask = os.umask(int(0o22))
```
Comment (galtsib) [inline client/tools/rhncfg/config_common/transactions.py]: oumask = os.umask(int('022', 8))

File: client/tools/rhncfg/config_common/transactions.py [hunk 8]
```diff
@@ -309,7 +309,7 @@ def deploy(self):
                         raise cfg_exceptions.DirectoryEntryIsFile(dirname)
                     if os.path.isdir(dirname):
                         s = os.stat(dirname)
-                        entry = { 'filemode': "%o" % (s[0] & 07777),
+                        entry = { 'filemode': "%o" % (s[0] & int(0o7777)),
```
Comment (galtsib) [inline client/tools/rhncfg/config_common/transactions.py]: not sure is it correct or not

File: client/tools/rhncfg/config_common/utils.py [hunk 9]
```diff
@@ -92,7 +98,7 @@ def mkdir_p(path, mode=None, symlinks=None, allfiles=None):
     made as a result
     """
     if mode is None:
-        mode = 0700
+        mode = int('0o700', 8)
```
Comment (galtsib) [inline client/tools/rhncfg/config_common/utils.py]: int('0700', 8)

File: client/tools/rhncfg/config_management/rhncfg_get.py [hunk 10]
```diff
@@ -92,28 +92,30 @@ def run(self):
...
             try:
                 dep_trans.deploy()
-            except Exception, e:
+            except Exception:
+                e = sys.exc_info()[1]
                 try:
                     dep_trans.rollback()
-                except FailedRollback, e2:
+                except FailedRollback:
+                    e2 = sys.exc_info()[1]
```
Comment (galtsib) [inline client/tools/rhncfg/config_management/rhncfg_get.py]: Look at line below: it doesn't work for Python 3. use our own function raise_with_tb

**Comment by galtsib** [comment]
please, fix also file actions/script.py

**Comment by galtsib** [comment]
config_management/rhncfg_get.py
line 114:
raise e2, "FAILED ROLLBACK:  ", sys.exc_info()[2]
line 118:
raise e, "Deploy failed, rollback successful:  ", sys.exc_info()[2]

**Comment by galtsib** [comment]
config_management/rhncfg_diff.py
line 143:
first_row = diff_output.next()
second_row = diff_output.next()

will not work for Python3: 'generator' object has no attribute 'next'

**Comment by galtsib** [comment]
config_common/rpc_wrapper.py
line 141:
msg = msg + "%s\n%s\n" % (sys.exc_type, sys.exc_value)

for Python 3: 
sys.exc_type = sys.exc_info()[0]
sys.exc_value = sys.exc_info()[1]

**Comment by galtsib** [comment]
config_common/local_config.py
line 117:
- return _get_config().keys()
+ return list(_get_config().keys())

**Comment by galtsib** [comment]
please use 2to3 just to check what it wants

**Comment by galtsib** [comment]
And, yes, you should rebase all your commits:
for example:
you have a commit b8cb40d9f6e28f0f2dfcb4357d08076b9ab8409f about fixing 'apply'. But actually, I see fixing of 'apply' in other commits. There are 2 commits about 'except', merge them into one. etc

**Comment by Pajinek** [comment]
I am not sure what is the best way to rewrite this code

`raise e2, "FAILED ROLLBACK: ", sys.exc_info()[2]`

My idea: modify function raise_with_tb(e) to raise_with_tb(e, msg=None)
https://github.com/spacewalkproject/spacewalk/blob/master/client/rhel/rhnlib/rhn/tb.py

what do you think about it?

File: client/tools/rhncfg/config_common/utils.py [hunk 11]
```diff
@@ -19,9 +19,15 @@
 import re
 import shutil
 import pwd
-import urlparse
```
Comment (galtsib) [inline client/tools/rhncfg/config_common/utils.py]: add import sys

File: client/tools/rhncfg/config_management/rhncfg_get.py [hunk 12]
```diff
@@ -92,30 +94,30 @@ def run(self):
...
             try:
                 dep_trans.deploy()
-            except Exception, e:
+            except Exception:
                 try:
                     dep_trans.rollback()
-                except FailedRollback, e2:
-                    raise e2, "FAILED ROLLBACK:  ", sys.exc_info()[2]
+                except FailedRollback:
+                    raise_with_tb(FailedRollback("FAILED ROLLBACK:  "))
```
Comment (galtsib) [inline client/tools/rhncfg/config_management/rhncfg_get.py]: please, add traceback sys.exc_info()[2]

**Comment by galtsib** [comment]
Merged...

**Pull request closed without merging**